core/commands/dag/dag.go

```diff
@@ -241,3 +278,421 @@ var DagResolveCmd = &cmds.Command{
+		api, _ = api.WithOptions(options.Api.Offline(true))
+
+		// grab a gc lock so that regardless of the size of the streamed-in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		// !!!NOTE!!! - We take .GCLock, *not* .PinLock. The way the current stack is
+		// built makes it impossible to take both a GC and a Pinning lock, regardless
+		// of orer. Doing so can potentially deadlock if another thread starts GC
+		// ( this is not theoretical - it's been reliably simulated with the insertion
+		// of strategic sleeps()s and hand-triggering GC )
+		//
+		// See api.Pin().Add() vs node.Pinning.Pin() comment further down
+		unlocker := node.Blockstore.GCLock()
+		defer func() { unlocker.Unlock() }()
+
+		doPinRoots, _ := req.Options[pinRootsOptionName].(bool)
+		showProgress, _ := req.Options[progressOptionName].(bool)
+
+		retCh := make(chan importResult, 1)
+		var counts ObjectCounts
+		go importWorker(req, &res, &api, &counts, retCh)
+
+		var progressTicker *time.Ticker
+		if showProgress {
+			progressTicker = time.NewTicker(500 * time.Millisecond)
+			defer progressTicker.Stop()
+		}
+
+		var roots *expectedRootsSeen
+
+	ImportLoop:
+		for {
+			select {
+			case ret := <-retCh:
+				if ret.err != nil {
+					return ret.err
+				}
+				roots = ret.roots
+				if progressTicker != nil {
+					progressTicker.Stop()
+					if err := res.Emit(&CarImportOutput{ObjectCounts: counts}); err != nil {
+						return err
+					}
+					os.Stderr.WriteString("\n")
+				}
+				break ImportLoop
+			case <-progressTicker.C:
+				if err := res.Emit(&CarImportOutput{ObjectCounts: counts}); err != nil {
+					return err
+				}
+			case <-req.Context.Done():
+				return req.Context.Err()
+			}
+		}
+
+		// opportunistic pinning: try whatever sticks
+		if doPinRoots {
+
+			for c, seen := range *roots {
+
+				// We need to re-retrieve a block, convert it to ipld, and feed it
+				// to the Pinning interface, sigh...
+				//
+				// If we didn't have the deadlocking-gclock/pinlock problem we could
+				// use the Api directly like so (though internally it does the same):
+				//
+				// // not ideal, but the pinning api takes only paths :(
+				// rp := path.NewResolvedPath(
+				// 	ipfspath.FromCid(c),
+				// 	c,
+				// 	c,
+				// 	"",
+				// )
+				//
+				// if err := api.Pin().Add(req.Context, rp, options.Pin.Recursive(true)); err != nil {
+
+				ret := RootMeta{Cid: &c, PresentInCar: seen}
+
+				if block, err := node.Blockstore.Get(c); err != nil {
+					ret.PinError = err
+				} else if nd, err := ipld.Decode(block); err != nil {
+					ret.PinError = err
+				} else if err := node.Pinning.Pin(req.Context, nd, true); err != nil {
+					ret.PinError = err
+				} else if err := node.Pinning.Flush(req.Context); err != nil {
+					ret.PinError = err
+				}
```

Comment:
failing to pin will set PinError, which is then emitted, but doesn't cause the overall command to return unsuccessfully. It may not be a fatal error - e.g. it seems okay to keep going through the other potential pinnable roots, but does seem like it may be worth reflecting in the command exit status.

Reply:
I will push some unrelated fixes for this, so it may show outdated shortly. But the reason the error is carried through is to be able to provide the following ( I am breaking one of the pins artificially here ) :
```
admin:~/devel/go-ipfs$ go run ./cmd/ipfs dag import --enc=json --progress=false lotus_testnet_256.car devnet.car 
{"Root":{"Cid":{"/":"bafy2bzaceaxm23epjsmh75yvzcecsrbavlmkcxnva66bkdebdcnyw3bjrc74u"},"PresentInCar":true,"PinErrorMsg":""}}
{"Root":{"Cid":{"/":"bafy2bzacebedrc4n2ac6cqdkhs7lmj5e4xiif3gu7nmoborihajxn3fav3vdq"},"PresentInCar":true,"PinErrorMsg":""}}
{"Root":{"Cid":{"/":"bafy2bzacede2hsme6hparlbr4g2x6pylj43olp4uihwjq3plqdjyrdhrv7cp4"},"PresentInCar":true,"PinErrorMsg":"not today"}}
{"Root":{"Cid":{"/":"bafy2bzaceb55n7uxyfaelplulk3ev2xz7gnq6crncf3ahnvu46hqqmpucizcw"},"PresentInCar":true,"PinErrorMsg":""}}
```
Your point of *NOT* exitting with 0 when this happens is valid though: @Stebalien how does one do that? Also what does this mean for http 4xx-what?

Reply:
To exit with a non-zero status code, return an error at the end. Something like "failed to recursively pin some roots".

In terms of the HTTP response, we indicate errors either by:

1. Returning 500.
2. Sending a error trailing HTTP header (if we discover the error after returning the first result.

Reply:
Done in https://github.com/ipfs/go-ipfs/pull/7011/commits/aa85f2efada25768a7b38f6162efe11be4aba25a